Harden @INC handling

demerphq opened the pull request:
This hardens `@INC` hook handling in a bunch of ways. It ensures error message show the truth, and it fixes segfaults, and other bugs, and it adds new facilities to INC hooks, ($SKIP and INCDIR)

Features (not bugfixes) in this PR:

- Add support for `$SKIP` this variable is set to the index of @INC that a hook exists in before the hook is executed. The hook may alter the value to set it to the integer predecessor of the index they want perl to look at next. Thus a hook that rewrites `@INC` can set `$SKIP` to -1 and have perl process the replaced contents. undef is supported to represent -1.
- Ensure that errors show the `@INC` entries inspected or called, and not the latest state of `@INC` which may not match at all.
- Add an INCDIR method, which is called just like an INC hook would be, but which can return a list of directories that perl will search. Returning zero directories is allowed.
- Add support for providing an object as the first argument of array style hook definitions. In this case we will pass the array in as the third argument so we can execute the call_method() with the object as the first argument.

Bugfixes:
- Fixes a segfault when an INC hook does something like this: `*INC=[ "foo", "bar", "baz"];`
- Error message being incorrect about what directories were inspected. For instance `push @INC, sub { @INC=() }`, will trigger right at the end of the @INC search, and the error message will say that we searched no directories, when in fact we searched the normal list.
- Allow blessed subrefs with no INC or INCDIR methods defined to be treated as unblessed coderefs.
- Allow blessed refs with no INC or INCDIR methods to be used in `@INC` so their overloads are triggered.

This should resolve all the conceptual issues raised in https://github.com/Perl/perl5/issues/20538

Edited: removed dedupe logic discussed below.

demerphq committed:
pp_ctl.c - rename vars to something more descriptive in require_file()

"i" and "ar" is a bit too minimal, and I will be refactoring
this code somewhat in following patches, so this preps it for
future changes.

demerphq committed:
pp_ctl.c - make ENTER debug data a bit more clear

We will have more than one INC hook in the future.

demerphq committed:
pp_ctl.c - refetch @INC from *INC after hook

The original value may have been freed by the time the hook
returns, so we have to refetch it immediately after execution.

We also move the declaration into a more minimal scope.

demerphq committed:
pp_ctl.c - eval INC hooks, and rethrow errors with more useful message

When an INC hook blows up debugging what is going on can be
somewhat difficult. This adds some debugging data if the error
message does not seem to be customized.

demerphq committed:
pp_ctl.c - require_file: truthful errors and tests: ties are called once

We need to keep track of what we actually checked. We cannot simply
report the state of @INC at the end of the require, as it might have
changed, possibly several times during the require. This also accounts
for most "silly" stuff that might upset our internal assumptions, for
instance where a tie might report one value to the code doing the
directory check and another in the error message.

We had long standing tests to see that @INC tie elements where called
"once" but they actually tested they were called twice despite claiming
otherwise. This fixes all of those test so that a tied @INC entry is
called exactly once, and whatever it returned the first time is placed
in the error message.

This includes a change to the require error message, so that where it
once said "@INC contains:" it now says "@INC entries checked:". Note
this patch requires parent v0.239 to be available (which was done
in the previous commit).

demerphq committed:
pp_ctl.c - support $INC in require_file @INC hook

$INC is localized to be the C level index of the loop over the @INC
array. At the end of the hook its value is assigned back to the C level
loop iterator (inc_idx). This allows a hook to control where in the @INC
array the loop should continue, for instance -1 represents "reprocess
from the beginning" (and as a convenience so does undef). This can be
useful if the @INC array is modified by a hook. Normally we would just
"continue along", but this may or may not be the right thing to do, so we
let the user decide.

demerphq committed:
pp_ctl.c - handle objects in @INC a bit more gracefully

If an object doesn't have an INC hook then don't call it. Either simply
stringify the ref (think overloads), OR, if it is a blessed coderef,
then just execute it like it was an unblessed coderef.

Also handle when an object is passed as the first argument of the
array form of call. Previously this would throw an exception as the
first argument on the stack when we call_method() would not be
blessed. When this is the scenario we pass in the array as the third
argument to the method.

demerphq committed:
pp_ctl.c - add support for an INCDIR hook

This hook returns a list of directories for Perl to search.
If it returns an empty list it acts like a no-op (except for the
error message). The return from INCDIR is always stringified,
they are not treated the same as normal @INC entries so no hooks
returning hooks.

demerphq committed:
pp_ctl.c - move logic closer to where it is used

No sense doing something that has not effect.

jkeenan commented:
> This hardens `@INC` hook handling in a bunch of ways. It ensures error message show the truth, and it fixes segfaults, and other bugs, and it adds new facilities to INC hooks, ($SKIP and INCDIR)
> 
> Features (not bugfixes) in this PR:
> 
[snip]
> 
> Bugfixes:
> 

Someone is bound to ask (so let's get this question out of the way first) ... 

This p.r. does a lot.  Could it be broken down into smaller p.r.'s?  For example, could the bugfixes be handled before the new features?

demerphq commented:
@kid51 i dont plan to break up this PR. I originally did this as a single patch. I took an extra day to break it up into a bunch of smaller commits. All of the logic involved is in a single call site, and basically interacts with each other. So i see it as all one thing, the @INC api is broken, segfaults, and is really easy to get very very wrong.  The patches here ameliorate those issues and provide solutions. For instance @ap said he can make the code in base.pm, which is fragile, not be fragile anymore with the features added here.  Breaking it up into separate PR's would just be more work for everyone involved, for no real benefit to the project.

demerphq committed:
t/ - INC hardening tests

These tests are pretty much all marked TODO. In the
following patches we will remove the TODO as we fix
the bug or implement the feature.

haarg commented:
I'm happy to see the fixes here, but have some concerns.

First, and most minor, the error when failing to load a module includes the string `@INC contains:`. With this patch and using INCDIR, it instead reports the directories it searched. This is, strictly speaking, incorrect. The extra directories are not and never were in `@INC`. This could be misleading because it makes it appear that the dynamically checked directories were part of the static search list.

Second, de-duplicating both directories and hook calls seems wrong. I'm not sure what type of issue only calling a given hook once is meant to protect against, and there isn't a test for it. And I can think if a number of reasons to have both duplicate hooks and duplicate directories in `@INC`. It seems like a needless limitation.

demerphq commented:
> First, and most minor, the error when failing to load a module includes the string @INC contains:. With this patch and using INCDIR, it instead reports the directories it searched. This is, strictly speaking, incorrect. The extra directories are not and never were in @INC. This could be misleading because it makes it appear that the dynamically checked directories were part of the static search list.

Yes. Indeed. I left the "`@INC contains`" part of the error message alone, as I feel like there is likely a large amount of code that looks for it in error messages. But at the same time, I feel pretty strongly that we should report where we looked, not what @INC looks like after we have looked. 

Previously we reported unreliable data. FOr instance:

```push @INC, sub { @INC= (); }```

Would cause the error message to show an empty list of directories searched, even though we have searched many. Similarly, a tied variable might return one thing to the code that does the directory lookup, but another when we build the error message.

This leaves us in a sticky place in terms of fixing this. Reporting the state of @INC after the call is not helpful if that data can be corrupted, and changing the constant text in the error message would mean that we have a lot more BBC fallout (and BBB fallout frankly - I am pretty confident B has code that looks for that string).

We either continue to have an ineffective and corruptable error message, or we change the text and break lots of stuff. I decided the least worse choice was to leave the constant text unchanged, but to show the correct data about what "things" were inspected.

I don't mind changing the constant text if people think that makes sense. I really do object to us producing error messages which are unhelpful or confusing however. 

> Second, de-duplicating both directories and hook calls seems wrong. I'm not sure what type of issue only calling a given hook once is meant to protect against, and there isn't a test for it.

I will have to double check that later, I think if you look carefully you will see that it is tested implicitly. 

> And I can think if a number of reasons to have both duplicate hooks and duplicate directories in @INC. It seems like a needless limitation.

I think you need to outline more on that. How would having the same directory search twice be a desirable and good thing? Are you thinking of a hook that writes to disk, and relies on the double listing to find it twice?  Seems far fetched to me, but ok, i guess. 

However as far as *hooks* go, I do not agree. I think calling the same hook more than once is far more likely to be an accident than it is deliberate, and the deduper does not prevent its deliberate use. It is meant to trap this:

```
push @INC, sub { unshift @INC, "."};
```
will infinite loop in a part of our code where we will eventually exhaust the stack and segfault. I dont think we should be able to write perl code that can segfault or infinite loop that easily. Another case would be this:

```
push @INC, sub { @INC= shuffle @INC };
```

Would we want that to shuffle @INC a random number of times?

It is also meant to make it easier to rewrite @INC and preserve your hook in the result, without having to guard against it getting called twice during the same require. Eg consider this:

```
push @INC, sub { @INC = grep { !ref($_) || ($_ != __SUB__); push @INC, @stuff, __SUB__; () }
```
With the deduper I can write the above and not worry about the callback getting called twice. It makes life nice and easy. But at the same time if someone really wants to create code that infinite loops like the above they can do this:

```
unshift @INC, [ sub { $INC[$INC]= [ $INC[$INC][0] ]; $INC-- } ];
```

I think its a prudent guard that will make life easier, and because of the array API does not make anything impossible which used to be possible. IOW, we do our best to keep people from blowing their foot off, but if they really want to do so they can.
 
If you can think of a number of reasons to have a hook called more than once can you please outline what they are, and explain why you couldn't use the array interface to hooks to achieve your result?

For instance, with the protection in place, if you wanted the same hook called at say the beginning and end of @INC you could do this:

```
$callback= sub { ... };
unshift @INC, [$callback];
push @INC, [$callback];
```

However it would stop this:

```
my $callback= [ sub { ... } ];
unshift @INC, $callback;
push @INC, $callback;
```

I don't think this is an unreasonable restriction. It makes it much harder to shoot yourself in the foot in a way that is super hard to debug, but it still allows power users to do what they want. Specifically it does NOT prevent anyone from creating a hook that gets called twice during the same require, if that is what they want, but it forces them to do it in a way where it must be deliberate and intentional, not accidental.

Note from the point of view of the internals:

```
$callback = sub { };
push @INC, [ $callback ], [ $callback ];
```
is two separate hooks, even though it resolves down to the same callback.

PS: Sorry for the lag, somehow I didnt see any notification that you had responded to this.

haarg commented:
> Yes. Indeed. I left the "`@INC contains`" part of the error message alone, as I feel like there is likely a large amount of code that looks for it in error messages. But at the same time, I feel pretty strongly that we should report where we looked, not what `@INC` looks like after we have looked.
> 
> ...
> 
> This leaves us in a sticky place in terms of fixing this. Reporting the state of `@INC` after the call is not helpful if that data can be corrupted, and changing the constant text in the error message would mean that we have a lot more BBC fallout (and BBB fallout frankly - I am pretty confident B has code that looks for that string).

I think we probably have to bite the bullet on that breakage. Showing the list of items actually searched seems like a decent change, but using an incorrect error doesn't seem viable to me. A quick scan over CPAN doesn't show too many things that are likely to be broken by this.


> I think you need to outline more on that. How would having the same directory search twice be a desirable and good thing? Are you thinking of a hook that writes to disk, and relies on the double listing to find it twice?

This is pretty much exactly the case I was thinking of.

> However as far as _hooks_ go, I do not agree. I think calling the same hook more than once is far more likely to be an accident than it is deliberate, and the deduper does not prevent its deliberate use. It is meant to trap this:
> 
> ```
> push @INC, sub { unshift @INC, "."};
> ```
> 
> will infinite loop in a part of our code where we will eventually exhaust the stack and segfault.

Does it actually do that? It leads to an infinite loop, but there's no recursion. It doesn't seem like it would exhaust the stack.

> I dont think we should be able to write perl code that can segfault or infinite loop that easily

There are pretty much infinite ways to create an infinite loop in perl. I don't see the value in trying to prevent this type, unless they lead to things like segfaults. Especially since `@INC` hooks are not common, and problems with them will usually show up at process start.

> Another case would be this:
> 
> ```
> push @INC, sub { @INC= shuffle @INC };
> ```
> 
> Would we want that to shuffle `@INC` a random number of times?

Honestly, yes. This really seems like a "you get to keep both pieces" problem. If someone does something that silly, we don't need to protect them.

As for a useful case for calling a hook multiple times, this PR provides a feature that would make them useful while debugging:

```
my $hook = sub { warn "reached $INC"; return; };
unshift @INC, $hook;
push @INC, $hook;
splice @INC, 5, 0, $hook;
```

> For instance, with the protection in place, if you wanted the same hook called at say the beginning and end of `@INC` you could do this:
> 
> ```
> $callback= sub { ... };
> unshift @INC, [$callback];
> push @INC, [$callback];
> ```
> 
> However it would stop this:
> 
> ```
> my $callback= [ sub { ... } ];
> unshift @INC, $callback;
> push @INC, $callback;
> ```
> 

This just feels like something that is confusing to document and not at all an obvious restriction to me.

ap commented:
> > Are you thinking of a hook that writes to disk, and relies on the double listing to find it twice?
> 
> This is pretty much exactly the case I was thinking of.

FWIW it is also the first thing that came to mind when I thought about plausible examples where it might be useful. 🙂

demerphq commented:
@haarg and @ap:

> I think we probably have to bite the bullet on that breakage.

Ok. What I did was a punt, I dont mind changing it to something like "`@INC` entries checked:" although suggestions welcome. Just so long as you take some of the heat if it breaks stuff. :-)

> There are pretty much infinite ways to create an infinite loop in perl. I don't see the value in trying to prevent this type, unless they lead to things like segfaults.

I have a pretty strong view on this. Maybe I am wrong about the segfault, it is possible I am confusing different bugs I encountered. However the infinite loop on something that should be a bounded process, combined with the action at a distance nature and the fact that allowing dupe calls makes it harder to write subs that rewrite `@INC` says to me that the block is sensible. I think its pretty reasonable to make it difficult to casually write inc bombs.

The example you gave is IMO perfectly served by wrapping the subs in fresh arrays.  I didnt find documenting the trick with arrays to be at all difficult, I have already done so in the PR.  Check the wording, it is pretty straight forward. https://github.com/Perl/perl5/pull/20547/commits/547f010f39d78b59e6342c0c96b2562106d4c198#diff-230e04648f168b0b966360f70f52dadc6e058290f63cb249d8539d43432112f6R6901

> This is pretty much exactly the case I was thinking of.
>>FWIW it is also the first thing that came to mind when I thought about plausible examples where it might be useful. slightly_smiling_face

I don't really see why a hook that does this would not just write the file to disk and then return a filehandle to the file it just created via the standard INC machinery. Why depend on the second directory trick? It starts to sound convoluted and contrived.

I feel much less strongly about deduping directories compared to my view on hooks, but it also seems totally reasonable to dedup at the app level. What about slow disk scenarios?

Anyway, since the deduping is the only controversial part of the PR I will separate out that part to a separate PR (at the same time changing the error message) and we can debate the deduping separately.  We can get the stuff we do agree on out of the way first.

ap commented:
> I don't really see why a hook that does this would not just write the file to disk and then return a filehandle to the file it just created via the standard INC machinery. Why depend on the second directory trick?

If you happen to have an open handle to the file already, sure, but maybe you called an external program to extract an archive or check out from a repo, or maybe you used a module that doesn’t expose the guts of the extraction process in this way, or whatever.

Basically whatever answer you give for “hooks can already return handles, what’s the point of `INCDIR`?” is also going to be an answer to your own question.

> Maybe I am wrong about the segfault, it is possible I am confusing different bugs I encountered.

FWIW I am entirely in agreement that nothing code does to `@INC` should cause a segfault, even as it’s being iterated. But I don’t see the point of trying to prevent an infinite loop here. For hooks, I get what would compel you, but even then I think it’s not just unnecessary but an active obstacle to people who might have use cases you aren’t imagining right now. I don’t like that kind of restriction; `@INC` hooks already aren’t everyday stuff so it’s not hard to imagine that someone having a weird use case you or I aren’t going to be able to come up with on the spot in this thread.

(Certainly “I need to record the stack depth so I can tell the top-level `require` from nested ones” is not a solution that just popped into my head at day 1 of working on the base.pm patch.)

Generally for this sort of thing I have a strong preference to operational models that are primitive (“`@INC` will be iterated by index”) but robust (“no segfaults, no matter what you try”). Those are easiest to use in unforeseen ways. The more restrictions you add to the model, the harder that gets.

(Note that `$INC` is an extension to this model which makes it both *easier*, *and* more robust, to employ in novel ways. ++ all around to that.)

I *am* glad to see you added an escape hatch to allow duplicate uses of the same hook.

And making mistakes difficult to make by default is generally also a principle I subscribe to. Maybe I’m not giving this enough weight.

But my subjective sense is that this may be a case of overcorrection – something I see where people react to a problem by solving it from multiple angles at once, when just one of those solutions would have perfectly sufficiently balanced out the overall situation by taking away knock-on effects, where the extra solutions create their own undesirable knock-on effects instead. Maybe I’m wrong, as I said, this is just my subjective sense. But I’d make that case to try one thing at a time and see if that suffices, because it will be easier to judge its necessity this way vs doing everything at once and trying to discern whether it was too much.

Maybe de-duping hooks by default to prevent mistakes but allowing duplicate calls if explicitly requested will turn out to be the correct balance. My bias at this time is of course that it will prove unnecessary – but I’m open to finding out otherwise.

ap commented:
> Anyway, since the deduping is the only controversial part of the PR I will separate out that part to a separate PR (at the same time changing the error message) and we can debate the deduping separately. We can get the stuff we do agree on out of the way first.

++

demerphq commented:
I have created https://github.com/Corion/parent/pull/12 as parent is one of the modules which depends on the test wording.

demerphq committed:
pod/ - document $INC and INCDIR

demerphq committed:
podcheck.t - make error message less confusing

podcheck.t assumes all non =head text will be indented
at least 4 spaces, but this wasn't explicit in the output
from podcheck for verbatim line length checks.

demerphq commented:
@haarg and @ap you ok with this now?

demerphq commented:
And the dedupe logic has been moved to https://github.com/Perl/perl5/pull/20588

haarg commented on pp_ctl.c:
```diff
@@ -4284,40 +4319,113 @@ S_require_file(pTHX_ SV *sv)
+                                }
+                                av_push(incdir_av, arg);
+                            }
+                            /* We copy $SIG into incdir_continue_inc_idx
```
Should this be `$INC`?

demerphq replied:
fixed

haarg commented on t/op/require_errors.t:
```diff
@@ -9,7 +9,12 @@ BEGIN {
+plan(tests => 71);
+
+
+# Dedupe @INC. As of 5.37.7 we won't inspect a directory twice, nor show a
```
This comment is out of date.

demerphq replied:
FIxed.

haarg commented on pp_ctl.c:
```diff
@@ -4284,40 +4319,113 @@ S_require_file(pTHX_ SV *sv)
+                            }
+                        }
+                        if (!method) {
+                            if (SvTYPE(SvRV(loader)) != SVt_PVCV) {
```
Should this allow `&{}` overloads?

demerphq replied:
Hrm. I dunno. On one hand, yeah from a completeness point of view we probably should. On the other hand, we don't support objects with overloaded `@{}` to implement the array API either, and in fact historically we would assume any object in @INC would have an INC method, and blindly call it as a method, with the resulting errors when it was a blessed sub with no methods. Its a tricky call. Prior to this patch you couldn't put an object in the array anyway, so nothing is going to break...

haarg replied:
If we only allow unblessed array refs, it feels like we should only allow unblessed code refs.

demerphq replied:
That doesn't make sense. If the array ref is blessed it is treated as an object, not as a hook wrapper. If it is unblessed then we treat it as a hook wrapper.  That is existing logic and can't change.

We used to ONLY allow subrefs in the zero slot of the wrapper, which made some sense because the only hook method we offered was INC. However we now have two methods, INCDIR and INC.  If we wish to have the wrapper approach work with both hooks we need to be able to supply an object in the zero slot.

The thread here doesnt show the full context. (I will update the comment below, it is misleading.)

```
                    if (sv_isobject(loader)) {
                        /* if it is an object and it has an INC method, then
                         * call the method. Otherwise assume its a blessed coderef.
                         */
                        HV *pkg = SvSTASH(SvRV(loader));
                        GV * gv = gv_fetchmethod_pvn_flags(pkg, "INC", 3, 0);
                        if (gv && isGV(gv)) {
                            method = "INC";
                        } else {
                            gv = gv_fetchmethod_pvn_flags(pkg, "INCDIR", 6, 0);
                            if (gv && isGV(gv)) {
                                method = "INCDIR";
                                is_incdir = TRUE;
                            }
                        }
                        if (!method) {
                            if (SvTYPE(SvRV(loader)) != SVt_PVCV) {
                                if (dirsv != loader)
                                    croak("Object with arguments in @INC does not support a hook method");
                                else
                                    goto treat_as_string;
                            }
                        }
                    }
```

So, if the loader is an object (either wrapper or unwrapped), we check to see if it has either hook methods. If it does not then we check if it is a blessed sub, and if it is then we call it, if it is not then we treat the object as a string.  In the special case where we have a wrapper (`dirsv != loader`), we are bit more careful, it doesn't make a lot of sense to use the wrapper api and include a non-coderef, non-hook bearing object. Unlike the unwrapped case, it doesn't make any sense to ignore the arguments (the wrapper), and fallback to strignifying the ref. Why would you provide a wrapper if it is just going to be thrown away to be stringified? 

In the old code, this would have worked differently. We would have seen the item was an object, tried to call the INC method and died with a cryptic error message.

With the new implementation, we expand things so that this:

```
push @INC, bless(sub { warn "@INC" }, "Me");
```

work the same as

```
push @INC, sub { warn "@INC" };
```

I believe that this is no bad thing. It helps debugging `@INC` hooks, and it doesn't break anything existing, and IMO it is more naturally DWIM anyway, the rules can say "An object with an INC or INCDIR hook, or a coderef". That isn't inconsistent with a blessed coderef with no method.

demerphq commented:
Ok, after discussion with haarg i am going ahead.
